.emacs.d/config.org
#+TITLE: Cort's Emacs Config
#+AUTHOR: Cort Spellman
#+DESCRIPTION: Cort's Emacs Config
#+STARTUP: showeverything
#+OPTIONS: toc:2
#+PROPERTY: header-args:emacs-lisp :lexical t

* TABLE OF CONTENTS :toc:
- [[#things-that-arent-working][Things That Aren't Working]]
  - [[#evil-want-c-i-jump-seems-to-have-no-effect][evil-want-C-i-jump seems to have no effect]]
- [[#todo][Todo]]
  - [[#show-lsp-and-other-errors][Show LSP And Other Errors]]
  - [[#keybinding-equivalents-of-mine-in-neovim][Keybinding Equivalents Of Mine In Neovim]]
  - [[#highlight-occurrences-of-term-under-cursor][Highlight Occurrences Of Term Under Cursor]]
- [[#elpaca-package-manager][Elpaca Package Manager]]
  - [[#install-and-enable-elpaca][Install And Enable Elpaca]]
  - [[#install-use-package-support][Install use-package Support]]
  - [[#have-use-package-always-install-packages][Have use-package Always Install Packages]]
  - [[#set-up-dired-to-use-gls-on-mac][Set Up dired To Use gls On Mac]]
- [[#set-path-as-in-terminals][Set Path As In Terminals]]
- [[#backup-files][Backup Files]]
- [[#lock-files][Lock Files]]
- [[#evil-mode][Evil Mode]]
- [[#font][Font]]
  - [[#set-the-font-face][Set The Font Face]]
- [[#general-keybindings][General Keybindings]]
- [[#git][Git]]
  - [[#magit][Magit]]
  - [[#git-time-machine][Git Time Machine]]
- [[#modeline][Modeline]]
  - [[#turn-on-column-numbers][Turn On Column Numbers]]
  - [[#use-doom-modeline][Use doom-modeline]]
- [[#org-mode][Org Mode]]
  - [[#indentation][Indentation]]
  - [[#table-of-contents][Table Of Contents]]
  - [[#enable-org-bullets][Enable Org Bullets]]
  - [[#create-blocks-in-org-files][Create Blocks In Org Files]]
- [[#projects][Projects]]
- [[#text][Text]]
  - [[#replace-selection-by-typing-in-insertish-mode-instead-of-appending][Replace Selection By Typing In Insertish Mode Instead Of Appending]]
  - [[#modify-word-characters][Modify Word Characters]]
  - [[#rainbow-parentheses][Rainbow Parentheses]]
  - [[#folding][Folding]]
  - [[#highlight-todo-and-friends][Highlight TODO And Friends]]
  - [[#set-background-color-of-color-code-to-that-color-code][Set Background Color Of Color-Code To That Color-Code]]
  - [[#insert-delimiters-as-pairs][Insert Delimiters As Pairs]]
  - [[#comment--uncomment-code][Comment / Uncomment Code]]
- [[#theme][Theme]]
- [[#ui][UI]]
  - [[#maximize-properly-on-macos][Maximize Properly on MacOS]]
  - [[#stop-cursor-from-blinking][Stop Cursor From Blinking]]
  - [[#turn-off-startup-message][Turn Off Startup Message]]
  - [[#disable-scrollbars][Disable Scrollbars]]
  - [[#disable-toolbar][Disable Toolbar]]
  - [[#display-line-numbers-in-certain-modes][Display Line Numbers In Certain Modes]]
  - [[#dont-truncate-lines][Don't Truncate Lines]]
  - [[#scroll-with-cursor-one-line-at-a-time][Scroll With Cursor One Line At A Time]]
  - [[#close-minibuffer-with-single-press-of-escape][Close Minibuffer With Single Press Of escape]]
  - [[#set-up-icons][Set Up Icons]]
  - [[#add-icons-to-dired][Add Icons To dired]]
  - [[#add-icons-to-ibuffer][Add Icons To ibuffer]]
  - [[#search][Search]]
  - [[#completion][Completion]]
  - [[#window-layout][Window Layout]]
- [[#undo][Undo]]
  - [[#undo-limits][Undo Limits]]
  - [[#undo-fu-for-linear-normal-and-logical-undo--redo][undo-fu For Linear (normal and logical) undo / redo]]
  - [[#preserve-undo-history-across-emacs-sessions][Preserve Undo History Across Emacs Sessions]]
  - [[#undo-tree-vundo][Undo Tree (vundo)]]
- [[#which-key][Which Key]]
- [[#languages][Languages]]
  - [[#lsp][LSP]]
  - [[#treesitter][Treesitter]]
  - [[#clojure][Clojure]]
  - [[#coffeescript][Coffeescript]]
  - [[#docker][Docker]]
  - [[#-elixir][;; Elixir]]
  - [[#go][Go]]
  - [[#lua][Lua]]
  - [[#markdown][Markdown]]
  - [[#-rust][;; Rust]]
  - [[#-svelte][;; Svelte]]
  - [[#yaml][YAML]]

* Things That Aren't Working
** evil-want-C-i-jump seems to have no effect
Workaround:
#+begin_src emacs-lisp
  (with-eval-after-load 'general
    (general-define-key
     :states '(motion)
     :keymaps 'override
     "C-i" 'evil-jump-forward))
#+end_src



* Todo
** Show LSP And Other Errors
*** I want the errors shown in the buffer.
*** I don't want to _have to_ mouse hover.
*** I don't want to _have to_ bring up a separate buffer to see errors.
** Keybinding Equivalents Of Mine In Neovim
** Highlight Occurrences Of Term Under Cursor
*** lsp-mode provides this but the functionality should be provided by treesitter when LSP isn't running. Also, I should be able to use * to search for the term under the cursor to see all such terms and the highlighting should persist until I clear it. (This provides a textual fallback.)



* Elpaca Package Manager
** Install And Enable Elpaca
#+begin_src emacs-lisp
  (defvar elpaca-installer-version 0.11)
  (defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
  (defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
  (defvar elpaca-repos-directory (expand-file-name "repos/" elpaca-directory))
  (defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
                                :ref nil :depth 1 :inherit ignore
                                :files (:defaults "elpaca-test.el" (:exclude "extensions"))
                                :build (:not elpaca--activate-package)))
  (let* ((repo  (expand-file-name "elpaca/" elpaca-repos-directory))
         (build (expand-file-name "elpaca/" elpaca-builds-directory))
         (order (cdr elpaca-order))
         (default-directory repo))
    (add-to-list 'load-path (if (file-exists-p build) build repo))
    (unless (file-exists-p repo)
      (make-directory repo t)
      (when (<= emacs-major-version 28) (require 'subr-x))
      (condition-case-unless-debug err
          (if-let* ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
                    ((zerop (apply #'call-process `("git" nil ,buffer t "clone"
                                                    ,@(when-let* ((depth (plist-get order :depth)))
                                                        (list (format "--depth=%d" depth) "--no-single-branch"))
                                                    ,(plist-get order :repo) ,repo))))
                    ((zerop (call-process "git" nil buffer t "checkout"
                                          (or (plist-get order :ref) "--"))))
                    (emacs (concat invocation-directory invocation-name))
                    ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
                                          "--eval" "(byte-recompile-directory \".\" 0 'force)")))
                    ((require 'elpaca))
                    ((elpaca-generate-autoloads "elpaca" repo)))
              (progn (message "%s" (buffer-string)) (kill-buffer buffer))
            (error "%s" (with-current-buffer buffer (buffer-string))))
        ((error) (warn "%s" err) (delete-directory repo 'recursive))))
    (unless (require 'elpaca-autoloads nil t)
      (require 'elpaca)
      (elpaca-generate-autoloads "elpaca" repo)
      (let ((load-source-file-function nil)) (load "./elpaca-autoloads"))))
  (add-hook 'after-init-hook #'elpaca-process-queues)
  (elpaca `(,@elpaca-order))
#+end_src

** Install use-package Support
The `:init` keyword executes code before a package is loaded. It accepts one or more forms, up to the next keyword.

The `:config` keyword executes code after a package is loaded. In cases where loading is done lazily, this execution is deferred until after the autoload occurs.

#+begin_src emacs-lisp
  (elpaca elpaca-use-package
    ;; Enable use-package :ensure support for Elpaca.
    (elpaca-use-package-mode))

  ;; Turns off elpaca-use-package-mode current declaration
  ;; Note this will cause evaluate the declaration immediately. It is not deferred.
  ;; Useful for configuring built-in emacs features.
  (use-package emacs
    :ensure nil
    :init
    ;; Add prompt indicator to `completing-read-multiple'.
    ;; We display [CRM<separator>], e.g., [CRM,] if the separator is a comma.
    (defun crm-indicator (args)
      (cons (format "[CRM%s] %s"
                    (replace-regexp-in-string
                     "\\`\\[.*?]\\*\\|\\[.*?]\\*\\'" ""
                     crm-separator)
                    (car args))
            (cdr args)))
    (advice-add #'completing-read-multiple :filter-args #'crm-indicator)

    ;; Do not allow the cursor in the minibuffer prompt
    (setq minibuffer-prompt-properties
          '(read-only t cursor-intangible t face minibuffer-prompt))
    (add-hook 'minibuffer-setup-hook #'cursor-intangible-mode)

    :config
    (setq ring-bell-function #'ignore)
    ;; Revert buffers when the underlying file has changed
    (global-auto-revert-mode 1)
    ;; Revert Dired and other buffers
    (setq global-auto-revert-non-file-buffers t)

    :custom
    ;; Support opening new minibuffers from inside existing minibuffers.
    (enable-recursive-minibuffers t)

    ;; TAB cycle if there are only few candidates
    ;; (completion-cycle-threshold 3)

    ;; Enable indentation+completion using the TAB key.
    ;; `completion-at-point' is often bound to M-TAB.
    (tab-always-indent 'complete)

    ;; Emacs 30 and newer: Disable Ispell completion function.
    ;; Try `cape-dict' as an alternative.
    (text-mode-ispell-word-completion nil)

    ;; Emacs 28 and newer: Hide commands in M-x which do not work in the current
    ;; mode. Vertico commands are hidden in normal buffers. This setting is
    ;; useful beyond Vertico.
    (read-extended-command-predicate #'command-completion-default-include-p))
#+end_src

** Have use-package Always Install Packages
#+begin_src emacs-lisp
  (setq use-package-always-ensure t)
#+end_src


** Set Up dired To Use gls On Mac
The Mac built-in `ls` doesn't support the `--dired` option so we use `gls` from GNU coreutils.

#+begin_src emacs-lisp
  (when (string= system-type "darwin")
    (setq dired-use-ls-dired t
          insert-directory-program "gls"
          dired-listing-switches "-ahl"))
#+end_src



* Set Path As In Terminals
#+begin_src emacs-lisp
  (use-package exec-path-from-shell
    :demand t
    :after elpaca
    :config
    (when (memq window-system '(mac ns x))
      (message "Mac OS system; will call exec-path-from-shell-initialize")
      (exec-path-from-shell-initialize)
      (message "Did call exec-path-from-shell-initialize"))
    ;; (when (daemonp)
    ;;   (exec-path-from-shell-initialize))
    )
#+end_src



* Backup Files
Source: https://emacs.stackexchange.com/a/36

#+begin_src emacs-lisp
(let ((backup-dir "~/tmp/emacs/backups")
      (auto-saves-dir "~/tmp/emacs/auto-saves/"))
  (dolist (dir (list backup-dir auto-saves-dir))
    (when (not (file-directory-p dir))
      (make-directory dir t)))
  (setq backup-directory-alist `(("." . ,backup-dir))
        auto-save-file-name-transforms `((".*" ,auto-saves-dir t))
        auto-save-list-file-prefix (concat auto-saves-dir ".saves-")
        tramp-backup-directory-alist `((".*" . ,backup-dir))
        tramp-auto-save-directory auto-saves-dir))

(setq backup-by-copying t    ; Don't delink hardlinks
      delete-old-versions t  ; Clean up the backups
      version-control t      ; Use version numbers on backups,
      kept-new-versions 5    ; keep some new versions
      kept-old-versions 2)   ; and some old ones, too
#+end_src



* Lock Files
#+begin_src emacs-lisp
  (setq lock-file-name-transforms `((".*" "~/tmp/emacs-lockfiles/" t)))
#+end_src


* Evil Mode
#+begin_src emacs-lisp
  (use-package evil
    :demand t
    :init
    (setq evil-want-integration t
          evil-want-keybinding nil
          evil-want-C-i-jump t
          evil-want-Y-yank-to-eol t
          evil-vsplit-window-right t
          evil-split-window-below t
          ;; https://evil.readthedocs.io/en/latest/settings.html#elispobj-evil-respect-visual-line-mode
          evil-respect-visual-line-mode t
          evil-undo-system 'undo-fu)
    (evil-mode))

  (use-package evil-collection
    :after evil
    :config
    (setq evil-collection-mode-list '(dashboard
                                      dired
                                      eldoc
                                      elisp-mode
                                      elpaca
                                      git-timemachine
                                      ibuffer
                                      magit
                                      org
                                      vc-annotate
                                      which-key))
    (evil-collection-init))

  (use-package evil-tutor)
#+end_src



* Font
** Set The Font Face
#+begin_src emacs-lisp
  (set-face-attribute 'default nil
		      :font "Monaco"
		      :height 140
		      :weight 'medium)
  (set-face-attribute 'variable-pitch nil
		      :font "JetBrainsMono Nerd Font"
		      :height 140
		      :weight 'medium)
  (set-face-attribute 'fixed-pitch nil
		      :font "Monaco"
		      :height 140
		      :weight 'medium)

  ;; This sets the default font on all graphical frames created after restarting Emacs.
  ;; Does the same thing as 'set-face-attribute default' above, but emacsclient fonts
  ;; are not right unless I also add this method of setting the default font.
  (add-to-list 'default-frame-alist '(font . "Monaco-14"))

  ;; Uncomment the following line if line spacing needs adjusting.
  (setq-default line-spacing 0.12)
#+end_src



* General Keybindings
general.el is the keybinding-setting program

#+begin_src emacs-lisp
  (use-package general
    :demand t
    :after evil
    :config
    (general-evil-setup)

    ;; Set space = "SPC" as global leader
    (general-create-definer cs/leader-keys
      :states '(normal insert visual emacs)
      :keymaps 'override
      :prefix "SPC" ; set leader
      :global-prefix "M-SPC") ; access leader in insert mode

    (cs/leader-keys
      "c" '(:ignore t :wk "Emacs config")
      "c c" '((lambda () (interactive) (find-file "~/.emacs.d/config.org")) :wk "Open Emacs config")
      "c r" '((lambda ()
                (interactive)
                (load-file "~/.emacs.d/init.el")
                (ignore (elpaca-process-queues)))
              :wk "Reload Emacs config"))

    (cs/leader-keys
      "h" '(:ignore t :wk "Help")
      "h b" '(describe-bindings :wk "Describe bindings")
      "h c" '(describe-command :wk "Display command")
      "h C" '(describe-char :wk "Describe character")
      "h d" '(:ignore t :wk "Emacs documentation")
      "h d d" '(view-emacs-debugging :wk "Emacs debugging")
      "h d f" '(view-emacs-FAQ :wk "Emacs FAQ")
      "h d m" '(info-emacs-manual :wk "Emacs manual")
      "h d n" '(view-emacs-news :wk "Emacs news")
      "h d p" '(view-emacs-problems :wk "Emacs problems")
      "h e" '(view-echo-area-messages :wk "Echo area messages")
      "h f" '(describe-function :wk "Describe function")
      "h F" '(describe-face :wk "Describe face")
      "h i" '(info :wk "Info")
      "h k" '(describe-key :wk "Describe key")
      "h l" '(view-lossage :wk "Recent keystrokes")
      "h m" '(describe-mode :wk "Describe mode")
      "h v" '(describe-variable :wk "Describe variable")
      "h w" '(where-is :wk "Keybinding for command"))

    (cs/leader-keys
      "b" '(:ignore t :wk "Buffer")
      "b i" '(ibuffer :wk "Ibuffer")
      "b k" '(kill-this-buffer :wk "Kill this buffer")
      "b n" '(next-buffer :wk "Next buffer")
      "b p" '(previous-buffer :wk "Previous buffer")
      "b r" '(revert-buffer :wk "Reload buffer"))

    (cs/leader-keys
      "e" '(:ignore t :wk "Eval")
      "e b" '(eval-buffer :wk "Eval buffer")
      "e f" '(eval-defun :wk "Eval defun")
      "e e" '(eval-last-sexp :wk "Eval last sexp")
      "e r" '(eval-region :wk "Eval region")
      "e :" '(eval-expression :wk "Eval expression"))

    (general-define-key
     :states 'normal
     :keymaps 'override
     "-" 'dired-jump)

    ;; Defining these in the (default) global keymap in addition to in
    ;; evil states, below, makes them work in ihelp buffers.
    (general-define-key
     "C-h" 'evil-window-left
     "C-j" 'evil-window-down
     "C-k" 'evil-window-up
     "C-l" 'evil-window-right)

    (general-define-key
     :states '(normal insert visual emacs)
     :keymaps 'override
     "C-h" 'evil-window-left
     "C-j" 'evil-window-down
     "C-k" 'evil-window-up
     "C-l" 'evil-window-right)
   )
#+end_src



* Git
** Magit
#+begin_src emacs-lisp
    (use-package transient
      :demand t)

    (use-package magit
      :after (:all transient evil evil-collection general)
      :custom
      (magit-display-buffer-function #'magit-display-buffer-fullframe-status-v1)
      (magit-bury-buffer-function #'magit-restore-window-configuration)
      (magit-no-confirm '(trash))
      (magit-section-initial-visibility-alist '((untracked . show)))
      
      ;; https://doc.endlessparentheses.com/Fun/format-time-string
      ;; %Y is the year
      ;; %m is the numeric month
      ;; %d is the day of the month, zero-padded
      ;; %H is the hour on a 24-hour clock
      ;; %M is the minute
      (magit-log-margin '(t "%Y-%m-%d %H:%M" magit-log-margin-width t 18))

      :config
      (cs/leader-keys
        "g" '(:ignore t :wk "Git")
        "g g" '(magit-status :wk "Magit Status"))

      (general-define-key
       :states '(normal visual)
       :keymaps '(magit-mode-map magit-status-mode-map)
       "n" 'magit-section-forward
       "p" 'magit-section-backward)

      (add-hook 'git-commit-mode-hook (lambda () (when git-commit-mode
                                                   (evil-insert-state)))))

#+end_src

** Git Time Machine
#+begin_src emacs-lisp
  (use-package git-timemachine)
#+end_src



* Modeline
** Turn On Column Numbers
#+begin_src emacs-lisp
  (setq column-number-mode t)
#+end_src


** Use doom-modeline
#+begin_src emacs-lisp
  (use-package doom-modeline
    :after nerd-icons
    :init
    (doom-modeline-mode 1)
    (setq doom-modeline-buffer-encoding nil)
    (setq doom-modeline-buffer-file-name-style 'relative-from-project))
#+end_src



* Org Mode
** Indentation
Disable electric indent in org-mode.

#+begin_src emacs-lisp
  (add-hook 'org-mode-hook (lambda () (electric-indent-local-mode -1)))
#+end_src



** Table Of Contents
Enable table of contents

#+begin_src emacs-lisp
  (use-package toc-org
    :commands toc-org-enable
    :init
    (add-hook 'org-mode-hook 'toc-org-enable))
#+end_src

** Enable Org Bullets
org-bullets displays indented bullets in org outlines instead of sequences of asterisks.

#+begin_src emacs-lisp
  (add-hook 'org-mode-hook 'org-indent-mode)
  (use-package org-bullets)
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
#+end_src

** Create Blocks In Org Files
Org-tempo is not a separate package but a module of org-mode that can be enabled. Org-tempo allows for `<s` followed by `TAB` to expand to a `begin_src` tag.
Other expansions available include:

| Typing the below + TAB | Expands to ...                          |
|------------------------+-----------------------------------------|
| <a                     | '#+BEGIN_EXPORT ascii' … '#+END_EXPORT  |
| <c                     | '#+BEGIN_CENTER' … '#+END_CENTER'       |
| <C                     | '#+BEGIN_COMMENT' … '#+END_COMMENT'     |
| <e                     | '#+BEGIN_EXAMPLE' … '#+END_EXAMPLE'     |
| <E                     | '#+BEGIN_EXPORT' … '#+END_EXPORT'       |
| <h                     | '#+BEGIN_EXPORT html' … '#+END_EXPORT'  |
| <l                     | '#+BEGIN_EXPORT latex' … '#+END_EXPORT' |
| <q                     | '#+BEGIN_QUOTE' … '#+END_QUOTE'         |
| <s                     | '#+BEGIN_SRC' … '#+END_SRC'             |
| <v                     | '#+BEGIN_VERSE' … '#+END_VERSE'         |

#+begin_src emacs-lisp
  (require 'org-tempo)
#+end_src


Prevents `<` from auto-pairing when electric-pair-mode is on.
Otherwise, org-tempo is broken when you try to type the above shortcuts.

#+begin_src emacs-lisp
  (add-hook 'org-mode-hook (lambda ()
                             (setq-local electric-pair-inhibit-predicate
                                         `(lambda (c)
                                            (if (char-equal c ?<) t (,electric-pair-inhibit-predicate c))))))
#+end_src



* Projects
Tools to navigate projects and files within them.

#+begin_src emacs-lisp
  (use-package projectile
    :demand t
    :after general
    :init
    (setq projectile-project-search-path '("~/Projects"))

    :config
    (projectile-mode +1)
    (cs/leader-keys
      "p" '(projectile-command-map :wk "Projectile")))
#+end_src



* Text
** Replace Selection By Typing In Insertish Mode Instead Of Appending
#+begin_src emacs-lisp
  (delete-selection-mode 1)
#+end_src

** Modify Word Characters
#+begin_src emacs-lisp
  (defun consider-underscore-word-character ()
    (modify-syntax-entry ?_ "w"))


.config/doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
;; (setq user-full-name "John Doe"
;;       user-mail-address "john@doe.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-symbol-font' -- for symbols
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
(setq doom-font (font-spec :family "Monaco" :size 13 :weight 'medium)
      doom-variable-pitch-font (font-spec :family "Fira Sans") ; inherits `doom-font''s :size
      doom-symbol-font (font-spec :family "JetBrainsMono Nerd Font" :size 13)
      doom-big-font (font-spec :family "Fira Mono" :size 19))

(setq-default line-spacing 0.12)

;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-one-light)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")


;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.



;; =============================================================================
;; Undo
;; =============================================================================
;; This gets its own section because it's very important and I've struggled with
;; buggy undo setups in the past.

;; Undo Limits
(setq undo-limit 67108864) ; 64mb.
(setq undo-strong-limit 100663296) ; 96mb.
(setq undo-outer-limit 1006632960) ; 960mb.

;; By default Doom Emacs uses undo-fu For Linear (normal and logical) undo / redo
;; We add undo tree with vundo
(use-package! vundo
  :after undo-fu)

;; Enable the undelete-frame function to reopen any of the last 16 closed
;; frames.
;; That means that I open something in a new frame, close it, and then change my
;; mind and want it back, I can simply call (undelete-frame).
;; Ex:
;; `C-u 2 M-x undelete-frame` calls `undelete-frame` with `arg` bound to 2 to
;; undelete the second-to-last deleted frame.
(undelete-frame-mode 1)



;; Close Minibuffer With Single Press Of escape
;; By default, Emacs requires pressing "ESC" three times to escape-quit the minibuffer. Change this to one:
;; (global-set-key [escape] 'keyboard-escape-quit)
(map! [escape] 'keyboard-escape-quit)

;; Enable word-wrap (almost) everywhere. This doesn't seem great.
;; (+global-word-wrap-mode 1)
;; This is the setting I used in my non-Doom config and it seemed to work.
(global-visual-line-mode 1)

;; Scroll With Cursor One Line At A Time
;; Instead of the default of half a screen at a time.
;; Note that 0 is the default, which makes Emacs scroll half a screen when point goes off-screen.
(setq scroll-step 1
      scroll-conservatively 10000)

;; Revert Dired and other buffers
(setq global-auto-revert-non-file-buffers t)


(setq doom-localleader-key ",")
(setq doom-localleader-alt-key "M-,")

(use-package! evil-collection
  :after evil
  :config
  (setq evil-collection-mode-list '(dashboard
                                    dired
                                    eldoc
                                    elisp-mode
                                    git-timemachine
                                    ibuffer
                                    magit
                                    org
                                    vc-annotate
                                    which-key))
  (evil-collection-init))

;; Switch evil to emacs state for mode so I have normal keybindings.
(dolist (mode '(cider-inspector-mode
                xref--xref-buffer-mode))
  (add-to-list 'evil-emacs-state-modes mode))



(after! doom-ui
  (map! "s-n" 'make-frame)
  (map! :nvie "s-n" 'make-frame)

  ;; Doom remaps 'delete-frame to 'doom/delete-frame-with-prompt so we undo that setting.
  (global-set-key [remap delete-frame] nil)
  ;; And then we enable closing windows using a conventional shortcut: `cmd/super+w`.
  (map! "s-w" 'delete-frame)
  (map! :nvie "s-w" 'delete-frame))

(map! :n "-" 'dired-jump)

;; Make org-mode links open in new frames.
(after! org-mode
  (setcdr (assq 'file org-link-frame-setup) 'find-file-other-frame))

;; Display line numbers in certain modes.
(add-hook! '(bash-mode-hook
             bash-ts-mode-hook
             emacs-lisp-mode
             org-mode-hook
             prog-mode-hook
             sh-mode-hook
             yaml-ts-mode-hook)
  (lambda () (display-line-numbers-mode 1)))

;; Don't insert delimiters and quotes as pairs in certain modes.
(add-hook! '(org-mode-hook
             text-mode-hook)
  (lambda ()
    (electric-pair-local-mode 0)
    (smartparens-mode -1)))

(use-package! hl-todo
  :config
  (setq hl-todo--regexp "\\(\\<\\(TODO\\|FIXME\\|HACK\\|REVIEW\\|NOTE\\|DEPRECATED\\|QUESTION\\)\\>\\)"
        hl-todo-keyword-faces `(("TODO"       warning bold)
                                ("FIXME"      error bold)
                                ("HACK"       font-lock-constant-face bold)
                                ("REVIEW"     font-lock-keyword-face bold)
                                ("NOTE"       success bold)
                                ("DEPRECATED" font-lock-doc-face bold)
                                ("QUESTION" font-lock-constant-face bold)))

  :hook
  (org-mode . hl-todo-mode)
  (prog-mode . hl-todo-mode))

(use-package! idle-highlight-mode
  :hook ((prog-mode emacs-lisp-mode lisp-mode) . idle-highlight-mode)
  :config (setq idle-highlight-idle-time 0.15))

;; Doom seems to add 'dired-omit-mode to 'dired-mode-hook. I don't want
;; dired-omit-mode; it hides various files in dired. I want to see what's there.
(after! dired-x
  (remove-hook 'dired-mode-hook 'dired-omit-mode))


(after! corfu
  (setq global-corfu-modes '(prog-mode emacs-lisp-mode lisp-mode)))

;; ;; Defining these in the (default) global keymap in addition to in
;; ;; evil states, below, makes them work in ihelp buffers.
;; (map! "C-h" 'evil-window-left
;;       "C-j" 'evil-window-down
;;       "C-k" 'evil-window-up
;;       "C-l" 'evil-window-right)
;; (map! :nive
;;       "C-h" 'evil-window-left
;;       "C-j" 'evil-window-down
;;       "C-k" 'evil-window-up
;;       "C-l" 'evil-window-right)

;; Let `s` key have normal Vim behavior in normal mode; don't use it for evil-snipe package.
(remove-hook 'doom-first-input-hook #'evil-snipe-mode)

(map! :map emacs-lisp-mode-map
      :localleader
      :prefix ("e" . "eval")
      :desc "Eval expression" ":" 'eval-expression)

(setq magit-display-buffer-function #'magit-display-buffer-fullframe-status-v1
      magit-bury-buffer-function #'magit-restore-window-configuration
      magit-no-confirm '(trash)
      magit-section-initial-visibility-alist '((untracked . show))

      ;; https://doc.endlessparentheses.com/Fun/format-time-string
      ;; %Y is the year
      ;; %m is the numeric month
      ;; %d is the day of the month, zero-padded
      ;; %H is the hour on a 24-hour clock
      ;; %M is the minute
      magit-log-margin '(t "%Y-%m-%d %H:%M" magit-log-margin-width t 18))

(add-hook 'git-commit-mode-hook (lambda () (when git-commit-mode (evil-insert-state))))

(defun cs/consider-underscore-word-character ()
  (modify-syntax-entry ?_ "w"))

(defun cs/consider-hyphen-word-character ()
  (modify-syntax-entry ?- "w"))

(add-hook 'text-mode-hook #'cs/consider-underscore-word-character)
(add-hook 'prog-mode-hook #'cs/consider-underscore-word-character)
(add-hook! '(emacs-lisp-mode-hook
             lisp-mode-hook
             clojure-mode-hook)
           #'cs/consider-underscore-word-character
           #'cs/consider-hyphen-word-character)

(add-hook! '(prog-mode-hook
             emacs-lisp-mode-hook
             lisp-mode-hook)
           #'rainbow-delimiters-mode)

(use-package! highlight-numbers
  :hook ((prog-mode conf-mode) . highlight-numbers-mode)
  :config (setq highlight-numbers-generic-regexp "\\_<[[:digit:]]+\\(?:\\.[0-9]*\\)?\\_>"))

(use-package! coffee-mode
  :after sourcemap
  :mode (("\\.coffee\\'" . coffee-mode))
  :config
  (add-hook 'coffee-mode-hook (lambda () (rainbow-delimiters-mode -1)))
  (setq coffee-args-compile '("-c" "-m"))
  (add-hook 'coffee-after-compile-hook
            (lambda (props)
              (sourcemap-goto-corresponding-point props)
              (delete-file (plist-get props :sourcemap)))))

(after! lsp-mode
  (add-hook 'lsp-mode-hook (lambda () (lsp-headerline-breadcrumb-mode 1)))

  (defun cs/lsp-ignore-semgrep-rulesRefreshed (_workspace notification)
    "Ignore semgrep/rulesRefreshed notification . It frequently pops up when
opening a buffer where lsp-mode is enabled. A quick google search says no one
knows why right now.
Source: https://emacs.stackexchange.com/a/82271"
    (when (equal (gethash "method" notification) "semgrep/rulesRefreshed")
      (lsp--info "Ignored semgrep/rulesRefreshed notification")
      t))

  (advice-add 'lsp--on-notification :before-until #'cs/lsp-ignore-semgrep-rulesRefreshed))

;; Check this out :)
;; (use-package! evil-lisp-state
;;   :init
;;   (setq evil-lisp-state-global t)
;;   :config
;;   (evil-lisp-state-leader "SPC k")
;;   (map! :map evil-lisp-state-map
;;         ";" (evil-lisp-state-enter-command wrap-comment))
;;   :custom
;;   (evil-lisp-state-enter-lisp-state-on-command nil))

;; set default cider repl so it doesn't prompt for the shell type
(setq cider-preferred-build-tool 'clojure-cli)
;; Consider immediate children of `(comment)` form to be top level forms so they
;; can be evaluated with eval-defun-* fns.
(setq clojure-toplevel-inside-comment-form t)

;; correct the ripgrep pcre support
(setq search-engine 'grep)

;; ignore the failing project tests
(setq cider-test-default-exclude-selectors '("system"))

;; automatically always load the dev and test profiles
(after! clojure-mode
  (setq cider-clojure-cli-parameters "-A:dev:test"))

(setq cider-eval-result-duration 'change)
